### PR TITLE
DT-174: Add health checks for elite2 and auth

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: helm_deploy/offender-case-notes/templates/deployment.yaml [hunk 1]
```diff
@@ -34,7 +34,7 @@ spec:
               protocol: TCP
           livenessProbe:
             httpGet:
-              path: /health
+              path: /ping
```
Comment (petergphillips) [inline helm_deploy/offender-case-notes/templates/deployment.yaml]: now using `ping` for liveness to stop cascading failures

File: src/main/java/uk/gov/justice/hmpps/casenotes/health/HealthCheck.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+public abstract class HealthCheck implements HealthIndicator {
+    private final RestTemplate restTemplate;
+
+    @Override
+    public Health health() {
+        try {
+            final var responseEntity = this.restTemplate.getForEntity("/ping", String.class);
+            return Health.up().withDetail("HttpStatus", responseEntity.getStatusCode()).build();
+        } catch (final RestClientException e) {
+            return Health.down(e).build();
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/hmpps/casenotes/health/HealthCheck.java]: any error takes health down

File: src/test/resources/application-test.yml [hunk 3]
```diff
@@ -18,10 +18,15 @@ spring:
...
-    enabled: false
+  shutdown.enabled: false
 
-app:
-  jwt-public-key: LS0tLS1CRUdJTiBQVUJMSUMgS0VZLS0tLS0NCk1JSUJJakFOQmdrcWhraUc5dzBCQVFFRkFBT0NBUThBTUlJQkNnS0NBUUVBc09QQXRzUUFEZGJSdS9FSDZMUDUNCkJNMS9tRjQwVkRCbjEyaEpTWFBQZDVXWUswSExZMjBWTTdBeHhSOW1uWUNGNlNvMVd0N2ZHTnFVeC9XeWVtQnANCklKTnJzLzdEendnM3V3aVF1Tmg0ektSK0VHeFdiTHdpM3l3N2xYUFV6eFV5QzV4dDg4ZS83dk8rbHoxb0NuaXoNCmpoNG14TkFtczZaWUY3cWZuaEpFOVd2V1B3TExrb2prWnUxSmR1c0xhVm93TjdHVEdOcE1FOGR6ZUprYW0wZ3ANCjRveEhRR2hNTjg3SzZqcVgzY0V3TzZEdmhlbWc4d2hzOTZuelFsOG4yTEZ2QUsydXA5UHJyOUdpMkxGZ1R0N0sNCnFYQTA2a0M0S2d3MklSMWVGZ3pjQmxUT0V3bXpqcmU2NUhvTmFKQnI5dU5aelY1c0lMUE1jenpoUWovZk1oejMNCi9RSURBUUFCDQotLS0tLUVORCBQVUJMSUMgS0VZLS0tLS0=
-  elite2-api-base-url: http://localhost:8999
-  oauth-api-rest-url: http://localhost:8998/auth/api
+management.endpoint:
+  health.cache.time-to-live: 0
+  info.cache.time-to-live: 0
```
Comment (petergphillips) [inline src/test/resources/application-test.yml]: need to turn off caching for integration tests

File: src/test/java/uk/gov/justice/hmpps/casenotes/health/HealthCheckTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import static net.javacrumbs.jsonunit.assertj.JsonAssertions.assertThatJson;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class HealthCheckTest extends ResourceTest {
+    @Test
+    public void testDown() {
+        elite2MockServer.subPing(404);
+        oauthMockServer.subPing(404);
+
+        final var response = testRestTemplate.getForEntity("/health", String.class);
```
Comment (JamieDraperUK) [inline src/test/java/uk/gov/justice/hmpps/casenotes/health/HealthCheckTest.java]: Might be much cleaner serializing the response to a map and asserting against properties as you did above.

File: src/test/java/uk/gov/justice/hmpps/casenotes/integration/wiremock/OauthMockServer.java [hunk 5]
```diff
@@ -16,12 +17,12 @@ public OauthMockServer() {
 
     public void subGetUserDetails(final AuthToken username) {
         stubFor(
-                WireMock.get(WireMock.urlPathMatching(API_PREFIX + "/user/"+username.name()))
-                        .willReturn(WireMock.aResponse()
+                get(urlPathMatching(String.format("%s/user/%s", API_PREFIX, username.name())))
+                        .willReturn(aResponse()
                                 .withHeader("Content-Type", "application/json")
                                 .withBody("{\n" +
```
Comment (JamieDraperUK) [inline src/test/java/uk/gov/justice/hmpps/casenotes/integration/wiremock/OauthMockServer.java]: Would it be cleaner to use map then just serialize to json?

File: src/test/java/uk/gov/justice/hmpps/casenotes/health/HealthCheckTest.java (hunk 4 above)
Comment (petergphillips) [reply src/test/java/uk/gov/justice/hmpps/casenotes/health/HealthCheckTest.java]: not sure what you mean.  Above we serialise to a string and then use `JsonContent` to compare the contents of a file.  I'll merge this PR and then push a new version to see what you think

File: src/test/java/uk/gov/justice/hmpps/casenotes/integration/wiremock/OauthMockServer.java (hunk 5 above)
Comment (petergphillips) [reply src/test/java/uk/gov/justice/hmpps/casenotes/integration/wiremock/OauthMockServer.java]: this is a test server for mocking responses, so should match what the other application actually responds

**Pull request merged**